Cypress test. Canvas 3D functionality. Basic actions

dvkruchinin opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
Add Cypress tests.
Used for testing ```.pcd``` files are created using [PCL](https://github.com/PointCloudLibrary/pcl).
Since ```.pcd``` is an image and when interacting with it in the DOM, there are no changes to the classes/attributes of the element, so visual regression testing is used based on comparing images using the ```jimp``` library.

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+
+    function pressKeyPerspectiveDirections(key, screenshotNameBefore, screenshotNameAfter) {
+        cy.get('.cvat-canvas3d-perspective').trigger('mouseover').screenshot(screenshotNameBefore);
+        cy.get('body').type(`{alt}${key}`).wait(300); //Wait to change point cloud position
```
Shouldn't we use keydown here with corresponding keycode? (as in other tests)
Can we get rid of wait() function? It doesn't look reliable.

dvkruchinin replied:
You don't have to use this for type (). Since this function passes kyeCode. For example ```Details: "{ code: AltLeft, which: 18 }"```
Cypress executes the command too quickly, and without using ```wait()```, the points cloud changes do not have time to occur. In this case, there will be no percentage change in pixels when comparing images.

bsekachev replied:
> Cypress executes the command too quickly, and without using wait(), the points cloud changes do not have time to occur. In this case, there will be no percentage change in pixels when comparing images.

Can we wrap the following commands to setTimeout without exact timeout value?

Theoretically event ``cy.get('body').type(`{alt}${key}`)`` generates a macrotask which will be done after current stack execution. If we wrap the next instructions to setTimeout, it also generates a macrotask which will be done after the previous one. 
```
setTimeout(() => {
    cy.get('.cvat-canvas3d-perspective').screenshot(screenshotNameAfter);
    compareImages(`${screenshotNameBefore}.png`, `${screenshotNameAfter}.png`);
})
```

dvkruchinin replied:
In this case, I get an error.
```
Error:                       CypressError: The following error originated from your test code, not from Cypress.
  > Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise.
The command that returned the promise was:
  > `cy.screenshot()`
The cy command you invoked inside the promise was:
  > `cy.get()`
Because Cypress commands are already promise-like, you don't need to wrap them or return your own promise.
Cypress will resolve your command with whatever the final Cypress command yields.
The reason this is an error instead of a warning is because Cypress internally queues commands serially whereas Promises execute as soon as they are invoked. Attempting to reconcile this would prevent Cypress from ever resolving.
```
I will try to regenerate the ```.pcd``` files. Perhaps this will help to solve this problem. Since in this view, points cloud is far away in the image and minimal changes are not visible in the screenshot.

dvkruchinin replied:
To exclude the use of ```wait()```, I replaced it with an additional ```click()``` on the element. For additional time, change the orientation of points cloud.

bsekachev replied:
Tested it locally and I do not have any issues if I remove both wait() and click(). It is quite strange.

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+
+    function pressKeyArrowDirections(key, screenshotNameBefore, screenshotNameAfter) {
+        cy.get('.cvat-canvas3d-perspective').trigger('mouseover').screenshot(screenshotNameBefore);
+        cy.get('body').type(key).wait(300); //Wait to change point cloud position
```
The same question

dvkruchinin replied:
Answered above.

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+        cy.get(element)
+            .screenshot(screenshotNameBefore)
+            .trigger('wheel', { deltaY: deltaY })
+            .wait(300) // Wait change points cloud position
```
And here

dvkruchinin replied:
Answered above.

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+            cy.get('.cvat-workspace-selector-dropdown')
+                .not('.ant-select-dropdown-hidden')
+                .within(() => {
+                    cy.get('[title="Standard 3D"]').should('have.class', 'ant-select-item-option-active');
```
Let's not to use ant classes. We can add classes on our side if necessary. Can you do it or do you need any help?

dvkruchinin replied:
I agree that it is better not to use them. I can spend some time adding additional classes. But I'll probably need some help.

dvkruchinin replied:
To exclude the use of these classes, I reworked the attempts to click on the blocked menu items and check the selector text. It should not change at the same time.

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+        cy.get('.cvat-player-filename-wrapper').should('contain.text', expectedText);
+    }
+
+    function pressKeyPerspectiveDirections(key, screenshotNameBefore, screenshotNameAfter) {
```
The function name does not correspond to the things the function actually does.
I would suggest renaming it to ``testPerspectiveChangeOnKeyPress`` for example.

dvkruchinin replied:
Renamed

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+        compareImages(`${screenshotNameBefore}.png`, `${screenshotNameAfter}.png`);
+    }
+
+    function pressKeyArrowDirections(key, screenshotNameBefore, screenshotNameAfter) {
```
The same comment about the function name

dvkruchinin replied:
Renamed

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    function checkFilenameWrapperText(expectedText) {
+        cy.get('.cvat-player-filename-wrapper').should('contain.text', expectedText);
```
Do we really need this function that consist of one line?

dvkruchinin replied:
Function removed.

bsekachev commented on tests/cypress/integration/canvas3d_functionality/case_56_canvas3d_functionality_basic_actions.js:
```diff
@@ -0,0 +1,158 @@
+        compareImages(`${screenshotNameBefore}.png`, `${screenshotNameAfter}.png`);
+    }
+
+    function wheelDirection(element, deltaY, screenshotNameBefore, screenshotNameAfter) {
```
The same comment about the function name

dvkruchinin replied:
Renamed.

bsekachev reviewed (changes requested):
Will the tests for context images be written later?

dvkruchinin commented:
> Will the tests for context images be written later?

I can add this test, but it will have to be skipped. Because I never managed to display the context image. Maybe it's a bug. I was passing on information @vnishukov . 
For example if create a task use ```cvat/apps/engine/tests/assets/test_pointcloud_pcd.zip``` than a context image not displayed.
But displayed if use the ```cvat/apps/engine/tests/assets/test_velodyne_points.zip```.

bsekachev commented:
> > Will the tests for context images be written later?
> 
> I can add this test, but it will have to be skipped. Because I never managed to display the context image. Maybe it's a bug. I was passing on information @vnishukov .

Hi, @vnishukov . Have you discussed the issue on today's meeting?

@dvkruchinin In this case let's wait till it will be fixed.

bsekachev commented:
@dvkruchinin 

Could you please resolve my comments in this PR? After that we will merge it. Tests for context image are postponed.

bsekachev commented:
Locally I always get failed the latest scenario.
Do you have any thoughts?

![Screenshot from 2021-03-15 18-46-48](https://user-images.githubusercontent.com/40690378/111181241-d2b09380-85be-11eb-89e3-f7fd79ba14b7.png)

bsekachev commented:
After a lot of testing I have came around another stable error (the previous has gone and I have not changed code anyhow)

![Screenshot from 2021-03-15 20-58-58](https://user-images.githubusercontent.com/40690378/111199072-4b205000-85d1-11eb-84a7-70e4dc71d86d.png)

dvkruchinin commented:
Without waiting, I get an error comparing images only on the wheel. I think we can't do without wait() here, since cypress is waiting for the wheel to execute, but the image does not change points cloud immediately. And even at the time of creating the screenshot, points cloud is still moving. Although the wheel event is already over at this time.
I think it will be easier for us to meet to make a common decision.

dvkruchinin commented:
I've reworked it a bit ```.pcd``` files. Added points cloud to 100,000 points for a clearer display. I also replaced ```wait()``` for wheel with top/side/front for a loop. This gives a more reliable display of the image change.

bsekachev commented:
> I think we can't do without wait() here, since cypress is waiting for the wheel to execute, but the image does not change points cloud immediately

Cypress ``trigger()`` creates a synthetic event and it calls corresponding listeners (looks like it works in a sync way, try to perform simple example below in a browser console)

```js
window.document.addEventListener('wheel', () => { alert('wheel listener') }) // second
alert('a') // first
window.document.dispatchEvent(new WheelEvent('wheel'))
alert('b') // third
```

But according to the profiler, the browser performs requestAnimationFrame callbacks when IDLE to redraw canvas content and I do not know ways to perform any code after (setTimeout, requestIdleCallback do not work in Cypress tests). 

Cypress also says following:

> However, most of the time you don’t even have to worry about animations. Why not? Cypress will automatically wait for elements to stop animating prior to interacting with them via action commands like .click() or .type(). 
https://docs.cypress.io/faq/questions/using-cypress-faq.html#Some-of-my-elements-animate-in-how-do-I-work-around-that

So, itself it should work out with animations, but I am not sure if it only about css animations (transitions) or also about JS animations (requestAnimationFrame). I would make a suggestions that it is only about the first one. Probably asking community or developers is a good idea. Maybe them would help us with an idea how better to work with it.

They are just my thoughts after some investigation. Speaking about the PR, let's merge it now as is (I tried to run it several times and it worked). If it will be wrong later, we can just skip it and improve.
